test/graph_rewrites/assign_clusters.cc

```diff
@@ -0,0 +1,92 @@
+/*******************************************************************************
+ * Copyright 2017-2018 Intel Corporation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *******************************************************************************/
+#include "gtest/gtest.h"
+
+#include "ngraph_assign_clusters.h"
+#include "ngraph_utils.h"
+
+#include "tensorflow/core/graph/graph.h"
+#include "tensorflow/core/graph/node_builder.h"
+
+using namespace std;
+namespace ng = ngraph;
+
+namespace tensorflow {
+
+namespace ngraph_bridge {
+
+#define ASSERT_OK(x) ASSERT_EQ((x), ::tensorflow::Status::OK());
+
+// Given a graph of this form:
+//
+//  Node1--->Node2
```

Comment:
The cluster assignment happens after POST_REWRITE_FOR_EXEC. So tensorflow has already run its optimizations at this point.

Suppose we have a graph like this. N1 is a node with a static input marked by *
C1          C2
  \            /
    \        /
        +
        |
        |
        v*
        N1 <---------N2

Then after TF's optimizations, the graph should look like this after constant folding:
C(=C1+C2)
   |
   |
   v*
   N1<------N2

At which point we can handle this a static input


Here is a suggested test:
Have a constant-folding scenario and check that with L1 optimization enabled tensorflow, we get C and N1 in the same cluster. With L0 optimization on tensorflow, they should be on different clusters.

Reply:
> (Trying out ascii art) :-D

Nice, I like it  ;)

We could test this, but I think what you're describing here has more to do with properties of the constant-folding pass than with properties of our cluster-assignment pass. So unless we have reason to believe that there might be some failure emerging from a weird interaction between CF and cluster assignment, a simple test like:

```
C
|
|
v*
N1<------ N2
```

would be sufficient in and of itself, I think.